t/live.t

```diff
@@ -16,7 +16,11 @@ unless ($API_KEY) {
 
 my $future = DateTime->now + DateTime::Duration->new(years => 1);
 my $future_ymdhms = $future->ymd('-') . '-' . $future->hms('-');
-my $stripe = Net::Stripe->new(api_key => $API_KEY, debug => 1);
+my $stripe = Net::Stripe->new(
+    api_key     => $API_KEY,
+    api_version => $ENV{STRIPE_API_VERSION} || '',
```

Comment:
Rather than working off the user's version (over which the test has no control), I think a much better test would be to create multiple Stripe objects of different versions hard coded in the tests, and then test the expected change in behaviour.

Reply:
that sounds reasonable. i was mostly picking up on the approach used by @fheyer in the un-pulled #71.

so, what do you propose as far as restructuring the test script? an outer loop that iterates over a list of API versions running the entire set of tests, and conditionalizing arguments and expected return values where necessary?

equally important, how do you propose to determine the list of API versions to check? for example, in checking other API versions, i found that the current tests fail around test 159 for versions before [2012-10-26](https://stripe.com/docs/upgrades#2012-10-26). i am not proposing backward compatibility back that far. just pointing out that Stripe's indication of what is a major change might not be sufficient for determining what might break existing code.

Reply:
> so, what do you propose as far as restructuring the test script? an outer loop that iterates over a list of API versions running the entire set of tests, and conditionalizing arguments and expected return values where necessary?

Exactly.

> equally important, how do you propose to determine the list of API versions to check?

Definitely do it for all the versions it currently passes on. Apart from that, you choose the version(s) you want it to work on and add them in. If you do that in a way which people understand, we can all follow your footsteps when we want to add compatibility with a new version.

Does that make sense?

Reply:
Great to see progress here!

> Rather than working off the user's version (over which the test has no control), I think a much better test would be to create multiple Stripe objects of different versions hard coded in the tests, and then test the expected change in behaviour.

+1

Reply:
indeed it does make sense, as long as you recognize that "all the versions it currently passes on" is a fairly long list, making for a thorough, but long, testing process for end-users at install time. by my estimation, there are about 60 versions between current and 2012-10-26, where the tests fail.

also, history indicates that "in a way which people understand" is a tall order for me :-). for example, i prefer repeatable, dynamic approaches to this sort of thing, per command line idioms like i mentioned on #80. that kind of construct obviously makes it easy to generate the initial list, but it could be considered "fragile" in ways not intuitive to others.

so my starting point will probably be to use that idiom to generate a static qw() list, ordered from newest to oldest, leaving the onus on future maintainers to extend that list as newer versions become available.

or, is there a tasteful way to decouple what seems to me to be a divergent set of needs? ie, at install time, the end-user needs to test for either the a) default version to which their account is locked, or b) a specific version(s) to which their application is locked. and using a static list may still fall short for (a) in the case of an end-user with a new account whose default version post-dates the static list. and testing against that long, and growing, static list brings with it a significant "time burden" to the end-user at install time and contributes significant "pollution" of their test data.

maintainers on the other hand should definitely be testing all previously-working versions, and should be ensuring that the version list is up-to-date, so i don't see any of this as an undue burden on them.

now i hope i'm not being too finicky. :-) i'd rather hash these things out in conversation form, rather than making unnecessary commits and pull requests from misguided attempts.

thanks for your patience.

Reply:
> indeed it does make sense, as long as you recognize that "all the versions it currently passes on" is a fairly long list, making for a thorough, but long, testing process for end-users at install time.

Because they'd be talking to the Stripe API, these version dependent tests should be in t/live.t which doesn't run on installation, but only when `ENV{STRIPE_API_KEY}` is set.

> also, history indicates that "in a way which people understand" is a tall order for me :-). for example, i prefer repeatable, dynamic approaches to this sort of thing, per command line idioms like i mentioned on #80. that kind of construct obviously makes it easy to generate the initial list, but it could be considered "fragile" in ways not intuitive to others.

yes, it was fragile to me :-) but I get the gist of what you're saying. If it's too big, maybe hard coding into the live.t two versions - the earliest and the latest for which all tests pass, and then adding an --all parameter to the script which goes through all the versions we claim to work with?

Reply:
> Because they'd be talking to the Stripe API, these version dependent tests should be in t/live.t which doesn't run on installation, but only when `ENV{STRIPE_API_KEY}` is set.

aah, not sure i realized that, probably because i have had that value set during recent installs. so that i'm clear, t/live.t does run for end-users at install time, but it will short circuit if they don't have `ENV{STRIPE_API_KEY}` set? if that is the case, then i have much less concern about the length of the list of versions we are testing against.

> yes, it was fragile to me :-) but I get the gist of what you're saying. If it's too big, maybe hard coding into the live.t two versions - the earliest and the latest for which all tests pass, and then adding an --all parameter to the script which goes through all the versions we claim to work with?

pending your confirmation of the above, and your suggestion about the script parameter, i'm happy to configure a few different lists, \@all_versions, \@major_versions, \@broken_versions, for example, and allow the maintainers to select at run time. do you have a preference  for Getopt::Long vs other modules for parsing parameters passed to scripts?

Reply:
wonderful - and I'm happy with Getopt::Long